File: LeetCode/Medium/0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.java
class Solution {

    public boolean searchMatrix(int[][] matrix, int target) {
        for (int[] row : matrix) {
            if (search(row, target)) {
                return true;
            }
        }

        return false;
    }

    boolean search(int[] row, int target) {
        int left = 0;
        int right = row.length - 1;
        
        while (left <= right) {
            int mid = (left + right) >>> 1;
            int value = row[mid];

            if (value == target) {
                return true;
            }

            if (value < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return false;
    }

}

File: LeetCode/Medium/0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.kt
class Solution {

    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        var r = 0
        var c = matrix.first().lastIndex

        while (r < matrix.size && c >= 0) {
            if (matrix[r][c] == target) {
                return true
            }

            if (matrix[r][c] < target) {
                r++
            } else {
                c--
            }
        }
        
        return false
    }
}
